### PR TITLE
AMS-76: customize the view selector

### PR DESCRIPTION
**Pull request by grena**
**Description (for Contributor and Core Developer)**

This PR is about:
- giving extension points to the datagrid View Selector to be able to display other type of view entities
- changing the design of views (cf. screenshot)

screenshot from 2016-12-05 11 45 12
(_in the screenshot, the dropdown "views" is displayed **only** if there are at least 2 view types available_)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | N
| Added Behats                      | N
| Changelog updated                 |
| Review and 2 GTM                  |
| Micro Demo to the PO (Story only) |
| Migration script                  | N
| Tech Doc                          | N

### CODE REVIEW

**Comment by pierallard** [comment]
This test is failing on my PR
`features/datagrid/datagrid_views.feature:33`

I added `@unstable` tag to allow you to fix it in this PR.

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 1]
```diff
@@ -85,8 +91,23 @@ define(
...
+            },
+
+            /**
+             * Initialize the view type to display at initialization.
+             *
+             * @returns {Promise}
+             */
+            initializeViewTypes: function () {
+                this.currentViewType = 'view';
+                // TODO: IF PROJECT/VIEW ALREADY SELECTED, PICK THE RIGHT VIEW TYPE
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: This will be done in another PR, this PR is to display any view type only, not to select them.

Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: OK SOUNDS GOOD

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+             */
+            render: function () {
+                if (this.viewTypes.length > 1) {
+                    this.$el.html(this.template({
+                        __: __,
+                        currentViewType: this.currentViewType,
+                        viewTypes: this.viewTypes
+                    }));
+
+                    this.$('.AknActionButton').dropdown();
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js]: you should use a javascript class

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+            /**
+             * Method called on click on a view type in the dropdown of this template.
+             * It simply triggers an event with the selected view type.
+             *
+             * @param {Event} event
+             */
+            switchViewType: function (event) {
+                var viewType = $(event.target).data('value');
+                this.$('.AknActionButton').dropdown('toggle');
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js]: same

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 4]
```diff
@@ -63,6 +67,8 @@ define(
                     this.forwardMediatorEvents(module.config()['forwarded-events']);
                 }
 
+                this.viewTypes = module.config().view_types;
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: you should do this.config = module.config() it will be better for extensibility and more futureproof. Btw you could also use the form extension config and don't use requirejsconfig

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js (hunk 1 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: YEAH, AWESOME!!!!

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 5]
```diff
@@ -201,6 +240,21 @@ define(
...
+             * We need to re-trigger the select2 search event to fetch new views.
+             *
+             * @param {string} selectedType
+             */
+            switchViewType: function (selectedType) {
+                this.currentViewType = selectedType;
+
+                // Force the trigger of the search of select2
+                var searchTerm = this.$select2Instance.data('select2').search.val();
+                this.$select2Instance.select2('search', '');
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: please don't use $select2Instance, you can just use select2Instance

File: src/Pim/Bundle/UIBundle/Resources/public/less/libs/select2.less [hunk 6]
```diff
@@ -0,0 +1,194 @@
+@import '../base/variables';
+
+.grid-view-selector {
```
Comment (juliensnz) [inline src/Pim/Bundle/UIBundle/Resources/public/less/libs/select2.less]: As we accepted the new bem rules, shouldn't you follow them ?

Comment (willy-ahva) [inline src/Pim/Bundle/UIBundle/Resources/public/less/libs/select2.less]: According to @pierallard (tell me if i'm wrong) Select2.less will be trashed later.

File: src/Pim/Bundle/UIBundle/Resources/public/less/libs/select2.less [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+        right: 6px;
+      }
+
+      .AknActionButton {
+        height: 38px;
+        border-radius: 0;
+        border: 1px solid #aaa;
+        width: 107px;
+        font-weight: bold;
+      }
```
Comment (grena) [inline src/Pim/Bundle/UIBundle/Resources/public/less/libs/select2.less]: duplicate css for these components, don't use Akn stuff
(call mixin)

File: src/Pim/Bundle/UIBundle/Resources/public/less/libs/select2.less (hunk 6 above)
Comment (grena) [inline src/Pim/Bundle/UIBundle/Resources/public/less/libs/select2.less]: .select2.grid-view-selector

File: features/Pim/Behat/Decorator/Field/Select2Decorator.php [hunk 8]
```diff
@@ -89,7 +89,9 @@ public function open()
             $openerElement = $this->find('css', '.select2-search-field');
         }
 
-        $openerElement->click();
+        if (strpos($this->element->getAttribute('class'), 'select2-dropdown-open') === false) {
```
Comment (pierallard) [inline features/Pim/Behat/Decorator/Field/Select2Decorator.php]: if $this->element->hasClass('select2-dropdown-open')

File: features/datagrid/datagrid_views.feature [hunk 9]
```diff
@@ -29,14 +29,14 @@ Feature: Datagrid views
...
-  @unstable # This failing test will be fixed with http://github.com/akeneo/pim-community-dev/pull/5228
   Scenario: Successfully apply a view
     Given I am on the products page
     And I filter by "family" with operator "in list" and value "Boots"
     Then I should see product black-boots
     But I should not see products purple-sneakers and black-sneakers
     When I apply the "Default view" view
-    Then I should see products black-boots, purple-sneakers and black-sneakers
+    Then I should be on the products page
+    And I should see products black-boots, purple-sneakers and black-sneakers
```
Comment (pierallard) [inline features/datagrid/datagrid_views.feature]: Thanks @grena

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js (hunk 1 above)
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: WHY DO YOU USE UPPERCASE?

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-line-remove.html [hunk 10]
```diff
@@ -1,3 +1,3 @@
-<button type="button" class="btn btn-small btn-default muted <%- hidden ? 'hide' : '' %>" data-action="prompt-deletion">
-    <i class="icon-trash"></i>
+<button type="button" class="AknButton AknButton--important AknButton--small <%- hidden ? 'AknButton--hidden' : '' %>" data-action="prompt-deletion">
```
Comment (pierallard) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-line-remove.html]: Please use AknIconButton if it's an icon button without text

File: src/Pim/Bundle/UIBundle/Resources/public/css/select2.less [hunk 11]
```diff
@@ -265,47 +263,6 @@ a.select2-choice {
...
-          background-color: @akeneoLight;
-        }
-
-        .action, .select-view-action-list {
-          display: inline-block;
-          width: 100%;
-        }
-      }
-    }
-    .create-carret {
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/css/select2.less]: carrrrrrrrret

File: src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+      .view-selector-type-dropdown {
+        position: absolute;
+        top: 6px;
+        right: 6px;
+      }
+
+      .view-selector-type-switcher {
+        height: 38px;
+        border-radius: 0;
+        border: 1px solid #aaa;
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less]: Please use @AknBorderColor for all the borders (even if it's not exactly the same than the @yulldesign 's maquette)

File: src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+        width: 107px;
+        font-weight: bold;
+      }
+
+      .icon-search {
+        position: absolute;
+        top: 16px;
+        left: 12px;
+        font-size: 16px;
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less]: Use a defined font size (@AknLittleIconFontSize e.g.)

File: src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        color: @AknGrey;
+      }
+
+      .select2-input {
+        background: #ffffff !important;
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less]: prefer named colors when it's possible ("white")

File: src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    .select2-results {
+      padding: 0;
+      margin: 8px 8px 8px 8px;
+
+      .select2-result {
+        border-left: 1px solid @AknBorderColor;
+        border-top: 1px solid @AknBorderColor;
+        border-right: 1px solid @AknBorderColor;
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less]: Little tip: you can use
border: solid @AknBorderColor;
border-width: 1px 1px 1px 0

(don't change your syntax, it's 100% OK)

File: src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+          .select2-result-label-view {
+            .view-label.view-label-current {
+              color: @AknPurple;
+              font-style: italic;
+              font-weight: bold;
+            }
+
+            .grid-view-selector-line-overlay {
+              display: none;
+              background-color:rgba(17, 50, 77, 0.8);
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less]: Please define a color in the variables if this can not be derived from the other ones (with darken/lighten).
If it's an "hover" color, please use the default hover to have the same everywhere.

File: src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+        display: flex;
+
+        .create-button {
+          border-top-right-radius: 0;
+          border-bottom-right-radius: 0;
+          border-top-left-radius: 2px;
+          border-bottom-left-radius: 2px;
+        }
+
+        .create-carret {
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/less/lib/select2.less]: carret ? Did you mean "Caret" ?

File: src/Pim/Bundle/UIBundle/Resources/public/lib/select2/select2.css [hunk 18]
```diff
@@ -245,7 +245,7 @@ Version: 3.4.1 Timestamp: Thu Jun 27 18:02:10 PDT 2013
 
 .select2-search input {
     width: 100%;
-    height: auto !important;
+    height: auto;
```
Comment (pierallard) [inline src/Pim/Bundle/UIBundle/Resources/public/lib/select2/select2.css]: :heart:

File: features/Pim/Behat/Decorator/Field/Select2Decorator.php [hunk 19]
```diff
@@ -186,11 +188,17 @@ public function getAvailableValues()
...
 
         $resultElements = $this->spin(function () use ($widget) {
-            return $widget->findAll('css', '.select2-result-label');
-        }, 'Cannot find any .select2-result-label element.');
+            return $widget->findAll('css', '.select2-result-label, .select2-no-results');
+        }, 'Cannot find any .select2-result-label nor select2-no-results element.');
+
+        // Maybe a "No matches found"
+        $firstResult = $resultElements[0];
+        $noMatchesFound = strpos($firstResult->getAttribute('class'), 'select2-no-results') !== false;
```
Comment (grena) [inline features/Pim/Behat/Decorator/Field/Select2Decorator.php]: early return if no matches foudn

File: features/Pim/Behat/Decorator/Field/Select2Decorator.php [hunk 20]
```diff
@@ -201,4 +209,25 @@ public function getAvailableValues()
...
+    public function search($text)
+    {
+        $widget = $this->getWidget();
+        $widgetClasses = '.' . str_replace(' ', '.', $widget->getAttribute('class'));
+
+        $text = trim($text);
+
+        $this->getSession()->executeScript(
+            sprintf(
+                '$(\'%s .select2-search input[type="text"]\').val(\'%s\').trigger(\'input\');',
```
Comment (grena) [inline features/Pim/Behat/Decorator/Field/Select2Decorator.php]: .val()
.trigger()

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/LocaleController.php [hunk 21]
```diff
@@ -51,6 +51,6 @@ public function indexAction()
         $filteredLocales = $this->collectionFilter->filterCollection($locales, 'pim.internal_api.locale.view');
         $normalizedLocales = $this->normalizer->normalize($filteredLocales, 'internal_api');
 
-        return new JsonResponse($normalizedLocales);
+        return new JsonResponse(array_values($normalizedLocales));
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/LocaleController.php]: to see if fixed on CE master (rebase)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+        Backbone,
+        BaseForm,
+        template
+    ) {
+        return BaseForm.extend({
+            template: _.template(template),
+            viewTypes: [],
+            currentViewType: '',
+            events: {
+                'click [data-action="switchViewType"]': 'switchViewType'
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js]: class not data-action

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    ],
+    function (
+        $,
+        _,
+        __,
+        Backbone,
+        BaseForm,
+        template
+    ) {
+        return BaseForm.extend({
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js]: Backbone.view

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-type-switcher.html [hunk 24]
```diff
@@ -0,0 +1,14 @@
+<div class="AknDropdown view-selector-type-dropdown btn-group">
+    <button class="AknActionButton view-selector-type-switcher" title="" data-toggle="dropdown">
+        <%- __('pim_datagrid.view_selector.' + currentViewType).toUpperCase() %> <i class="AknActionButton-caret AknCaret"></i>
+    </button>
+    <ul class="AknDropdown-menu">
+        <% viewTypes.forEach(function (type) { %>
+        <li>
+            <a href="javascript:void(0);" data-action="switchViewType" data-value="<%- type %>" class="AknDropdown-menuLink" title="<%- __('pim_datagrid.view_selector.' + type).toUpperCase() %>">
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-type-switcher.html]: button

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 25]
```diff
@@ -179,18 +209,34 @@ define(
...
                 }.bind(this));
 
                 var $menu = this.$('.select2-drop');
+                var $search = this.$('.select2-search');
+
+                $search.prepend($('<i class="icon-search"></i>'));
+
+                // If more than 1 view type, we display the view type switcher module
+                if (this.config.viewTypes.length > 1) {
+                    var typeSwitcher = new ViewSelectorTypeSwitcher(this.config.viewTypes);
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: drop the shit, set view selector as parent

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 26]
```diff
@@ -179,18 +209,34 @@ define(
...
+                    $search.append(typeSwitcher.render().$el);
+
+                    typeSwitcher.listenTo(
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: this.listenExtensions? (in the configure)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 27]
```diff
@@ -179,18 +209,34 @@ define(
...
 
                 var $menu = this.$('.select2-drop');
+                var $search = this.$('.select2-search');
+
+                $search.prepend($('<i class="icon-search"></i>'));
+
+                // If more than 1 view type, we display the view type switcher module
+                if (this.config.viewTypes.length > 1) {
+                    var typeSwitcher = new ViewSelectorTypeSwitcher(this.config.viewTypes);
+                    $search.append(typeSwitcher.render().$el);
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: drop

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js [hunk 28]
```diff
@@ -0,0 +1,86 @@
...
+            /**
+             * Method called on click on a view type in the dropdown of this template.
+             * It simply triggers an event with the selected view type.
+             *
+             * @param {Event} event
+             */
+            switchViewType: function (event) {
+                var viewType = $(event.target).data('value');
+                this.$('.view-selector-type-switcher').dropdown('toggle');
+                this.setCurrentViewType(viewType);
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector-type-switcher.js]: drop since it will be set on event call

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 29]
```diff
@@ -133,9 +161,11 @@ define(
                             if (options.context && options.context.page) {
                                 page = options.context.page;
                             }
+
                             var searchParameters = this.getSelectSearchParameters(options.term, page);
+                            var viewFetcher = 'datagrid-' + this.currentViewType;
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: fetcher

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 30]
```diff
@@ -179,18 +209,34 @@ define(
...
+
+                // If more than 1 view type, we display the view type switcher module
+                if (this.config.viewTypes.length > 1) {
+                    var typeSwitcher = new ViewSelectorTypeSwitcher(this.config.viewTypes);
+                    $search.append(typeSwitcher.render().$el);
+
+                    typeSwitcher.listenTo(
+                        typeSwitcher, 'grid:view-selector:switch-type', this.switchViewType.bind(this)
+                    );
+                    typeSwitcher.setCurrentViewType(this.currentViewType);
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: trigger event with the view type
(or merge)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js [hunk 31]
```diff
@@ -269,7 +330,7 @@ define(
              * @return {array}
              */
             ensureDefaultView: function (choices) {
-                if (null !== this.defaultUserView) {
+                if (null !== this.defaultUserView || this.currentViewType !== 'view') {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: yoda

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-type-switcher.html (hunk 24 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-type-switcher.html]: data action to class

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-type-switcher.html [hunk 32]
```diff
@@ -0,0 +1,14 @@
+<div class="AknDropdown view-selector-type-dropdown btn-group">
+    <button class="AknActionButton view-selector-type-switcher" title="" data-toggle="dropdown">
+        <%- __('pim_datagrid.view_selector.' + currentViewType).toUpperCase() %> <i class="AknActionButton-caret AknCaret"></i>
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/grid/view-selector-type-switcher.html]: case with CSS

File: src/Pim/Bundle/UIBundle/Resources/public/lib/select2/select2.css (hunk 18 above)
Comment (grena) [inline src/Pim/Bundle/UIBundle/Resources/public/lib/select2/select2.css]: <3

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js (hunk 25 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: no, I can't, the parent is a simple input=hidden, since it's a select2.......

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js (hunk 27 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: I can't, select2 so impossible to use form_extensions

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js (hunk 26 above)
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/grid/view-selector.js]: same, not an extension, I can't.
